trpcultivate_phenotypes/src/Plugin/Validators/ValidDataFile.php

```diff
@@ -0,0 +1,198 @@
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('entity_type.manager')
+    );
+  }
+
+  /**
+   * Perform validation of data file.
+   * Checks include:
+   *  - Parameter filename or file id is valid.
+   *  - Has Drupal File Id number assigned and can be loaded.
+   *  - Valid file extension and file mime type configured by the file importer instance.
+   *  - File exists and not empty.
+   *  - File can be opened.
+   *
+   * @param string $filename
+   *   The full path to a file within the file system (Absolute file path).
+   * @param integer $fid
+   *   [OPTIONAL] The unique identifier (fid) of a file that is managed by Drupal File System.
+   *
+   * @return array
+   *   An associative array with the following keys.
+   *    - case: a developer focused string describing the case checked.
+   *    - valid: either TRUE or FALSE depending on if the file is valid or not.
+   *    - failedItems: an array of "items" that failed to be used in the message to the user. This is an empty array if the file input was valid.
+   */
+  public function validateFile(string $filename, int|null $fid = NULL) {
+
+    // Parameter check, verify that the filename/file path is valid.
+    if (empty($filename) && is_null($fid)) {
+      return [
+        'case' => 'Filename is empty string',
+        'valid' => FALSE,
+        'failedItems' => ['filename' => '', 'fid' => $fid]
+      ];
+    }
+
+    // Parameter check, verify the file id number is not 0 or negative values.
+    if (!is_null($fid) && $fid <= 0) {
+      return [
+        'case' => 'Invalid file id number',
+        'valid' => FALSE,
+        'failedItems' => ['filename' => $filename, 'fid' => $fid]
+      ];
+    }
+
+    // Holds the file object when file is a managed file.
+    $file_object = NULL;
+
+    // Load file object.
+    if (is_numeric($fid) && $fid > 0) {
+      // The file input is integer value, the file id number. Load the file object by fid number.
+      $file_object = File::load($fid);
+    }
+    elseif ($filename) {
+      // The file input is a string value, a path to the file. Locate the file entity
+      // by uri and load the file object using the returned file id number that matched.
+      $file_entities = $this->service_EntityTypeManager
+        ->getStorage('file')
+        ->loadByProperties(['uri' => $filename]);
+
+      $file_entity = reset($file_entities);
+
+      if ($file_entity) {
+        $fid = $file_entity->get('fid')->value;
+        $file_object = File::load($fid);
+      }
+    }
+
+    // Check that the file input provided returned a file object.
+    if (is_null($file_object)) {
+      return [
+        'case' => 'Filename or file id failed to load a file object',
+        'valid' => FALSE,
+        'failedItems' => ['filename' => $filename, 'fid' => $fid]
+      ];
+    }
+
```

Comment:
We need another case here that checks if the filename provided as a paramter matches the one in the file object loaded via the fid.

Reply:
Loading a file via fid means that the filename is not provided. Please clarify (include the case title and failed items to return).

Reply:
I think that's the way we think of it but technically **you can provide both parameters to this function**. If you do provide both a filename and a fid, the fid is preferentially used since it is checked first in the logic. 

So the scenario is:
if 
 - you are in the first block (i.e loading via fid because it was supplied)
 - you can load the file object via the fid
 - the filename parameter is also non-empty

then
 - check the filename parameter and the file_object->filename match

@carolyncaron mentions this should actually throw an exception since it should only happen if the developer sets the parameters wrong and I think I agree.

Reply:
What will be the case title and failed items to return?

Reply:
If you throw an exception then you don't need a case/failedItems since you do not return anything. We're in gather if you want to chat about it further 💞

Reply:
What will be the exception message so I can test it?